### DISCUSSION TITLE
Bug? Two-Way Broadcast/Observe slows down very quickly on two City Hubs

### DISCUSSION THREAD

**Discussion by lebenski** [pybricks/support#1491] [General]
I'm unsure whether I should raise this as a bug but figured I'd start with a discussion just incase there is something I can change in my setup to make this work.  As a bit of a background I've spent a couple of days trying to debug a quickly degrading performance in my two-way broadcast/observe City Hub setup.  I boiled my code down to the absolute basics and it seems to me that for two hubs that are both broadcasting/observing to each other, the speed of execution quickly slows to a complete crawl.

I'm using Pyrbricks Code and two City Hubs.  Both hubs are disconnected from the computer and the programs are run by using the button on the hub to rule out the bluetooth connection to the PC causing any slowness.

To recreate..

Hub1.py
```
from pybricks.hubs import CityHub
from pybricks.parameters import Color
from pybricks.tools import wait

hub = CityHub()
hub = CityHub(broadcast_channel=9, observe_channels=[1])

while True:

    hub.ble.broadcast("hub1-turn-red")
    hub.light.on(Color.RED)
    
    while not hub.ble.observe(1)=="hub2-turn-red":
        pass
    
    hub.ble.broadcast("hub1-turn-orange")
    hub.light.on(Color.ORANGE)

    while not hub.ble.observe(1)=="hub2-turn-orange":
        pass

    hub.ble.broadcast("hub1-turn-green")
    hub.light.on(Color.GREEN)

    while not hub.ble.observe(1)=="hub2-turn-green":
        pass
    
    wait(5000)
```

Hub2.py
```
from pybricks.hubs import CityHub
from pybricks.parameters import Color
from pybricks.tools import wait

hub = CityHub()
hub = CityHub(broadcast_channel=1, observe_channels=[9])

while True:

    while not hub.ble.observe(9)=="hub1-turn-red":
        pass

    hub.light.on(Color.RED)
    hub.ble.broadcast("hub2-turn-red")

    while not hub.ble.observe(9)=="hub1-turn-orange":
        pass
    
    hub.ble.broadcast("hub2-turn-orange")
    hub.light.on(Color.ORANGE)

    while not hub.ble.observe(9)=="hub1-turn-green":
        pass

    hub.ble.broadcast("hub2-turn-green")
    hub.light.on(Color.GREEN)
```

As you can see it's a very simple program where Hub2 is simply copying the light pattern of Hub1, from Red to Orange to Green.  Then finally when Hub1 has received a message that Hub2 has turned Green, it waits for 5 seconds then starts the loop again.

I timed how long it takes between Hub 1 turning Red to Hub 2 turning Green and these are the first 8 loops ignoring the first which darts by in a second or so.  Timings are approximate.. I did this using a manual stopwatch as I'm unable to see debug output when running the code on the hubs, obviously.

Loop 1: 4.03s
Loop 2: 11.93s
Loop 3: 21.04s
Loop 4: 34.58s
Loop 5: 46.86s
Loop 6: 63.04s
Loop 7: 81.17s
Loop 8: 81.83s

So as you can see it takes no time to degrade to a complete crawl.  

Am I doing something wrong in this code?  Is this a known issue?  Should I raise it as a bug?

EDIT: I should add that I also tried both Hub1.py and Hub2.py on each piece of City Hub hardware to rule out one of them being faulty.  The same issue occurs.

**Comment by laurensvalk** [comment]
Which firmware are you using? You can see it in the bottom right.

If you aren't doing so already, could you try this with the latest beta version, installing the firmware from there? To do so click "Try New Beta Features" on the settings panel of the app. 

Thanks!

![image](https://github.com/pybricks/support/assets/12326241/a52c9a60-6534-4840-97b1-900eaf7b9d9c)

**Comment by lebenski** [comment]
That did the trick!  Thanks so much.. my setup is running so much quicker now :) I don't suppose you happen to know the Issue number of the bug that was causing this so that I can read about it?  No worries if not.  Thanks for your help again.

**Reply by laurensvalk** [reply]
I am just about to try your programs... :smile:

**Reply by laurensvalk** [reply]
I suppose this is still not terribly fast, but maybe about as good as it gets when combining broadcasting and observing on the same City Hub. Glad it's about 80x faster then :smile:

There were quite a few updates since the last stable firmware, almost all concerning Bluetooth drivers, so it's not a coincidence that this got better. One way communication is about instantaneous now. Depending on your application, maybe there's a way to benefit from that.

https://github.com/pybricks/pybricks-micropython/compare/v3.3.0...HEAD

**Reply by lebenski** [reply]
It's by no means blazing fast but the speed is now consistent and I can work with it so thank you.

I'm building a machine to sort Magic The Gathering cards!  Here is a video of it this morning with the new beta firmware.

https://www.dropbox.com/scl/fi/1wwjusj8eu7llxt9sd1xn/IMG_5578.MOV?rlkey=zwlzj40hsqncjpfw7gaj0ifax&dl=0

So the basic idea is that my pc knows the order of the cards in the dispenser.  Let's say it wants to put the next card in the blue bucket.  

1. PC sends a message to the controller hub over UART saying "move the train to the Blue bucket".  
2. The controller hub broadcasts this message to the train.
3. The train moves into position with the help of the controller hub + color sensor which broadcasts the color sensor data to the train.
4. There is a few second pause for the card dispensing to occur.
5. The train broadcasts back to the hub saying "I'm ready for the next card"
6. The hub confirms receipt of that message with a broadcast back to the train and requests the next color from the pc.
7. The pc sends the color back over UART and we're back to step 1

At step 4 I'm doing the final bit of fine movement with my hands to align the bucket properly with the dispenser but I have a second color sensor in the post that will automate this for me.  I'm then going to use a motor + linear actuator to press the button on the card dispenser to complete the automation.

The final machine will also include a scanner (+ image recognition software) that will place the cards into the card dispenser in the first place.  I'm also planning to have at least 7 buckets in total on the final machine.

If you have any suggestion on how to improve any of that I'd love to hear it.  I'm new to all of this but having lots of fun with it.. Pybricks really is an outstanding piece of software which makes amateur robotics very accessible.

**Reply by laurensvalk** [reply]
Very cool! Looking forward to seeing this develop.

Another idea for getting accurate train positioning with potentially less detrimental effects of delay is to give the train a position index, and let the train handle the problem of going there and just responding "OK" when done. Perhaps with position markers on the track, see [e.g. this demo](https://www.youtube.com/watch?v=kwnSbnlhOkA).

And potentially drive the locomotive wheels with another motor that has position/speed control.

**Reply by lebenski** [reply]
Thanks for that!  I did consider using track markers but figured that by having the markers on the carriages themselves and a static color sensor that I could easily increase or decrease the capacity of the train by adding or removing extra carriages.  It would certainly reduce the delay to have the color sensor on the train though so I totally appreciate why that could be a better idea.

I'm just using the default Train motor that comes with 60198.  Two of them in fact!  Are you referring to one of these?  https://docs.pybricks.com/en/stable/pupdevices/motor.html.  What would be the advantage of such a motor over the basic train DC motor?  That speed is consistent regardless of battery level?

**Reply by laurensvalk** [reply]
Yes, they will run at a controlled speed, even as you add more carriages.

Plus, you can gear them down to make it go as slow (and accurate) as you like. Here is [some inspiration](https://www.eurobricks.com/forum/index.php?/forums/topic/189566-small-powered-up-shunter-with-pybricks-control/) but there should be more like these.

With a bit more software effort you can also speed and position control with the plain train motor if you count the sleepers, see https://pybricks.com/projects/sets/city/60198-cargo-train/speed-control/

**Reply by laurensvalk** [reply]
Actually, that reminds me. Maybe you can get the best of both worlds quite easily.

With this new beta you're using, you can broadcast `None` to stop broadcasting (this will properly stop the radio). So you could do that before you start listening for position/color info and get results super quickly. Then broadcast only for the `"I'm ready for the next card"` step.

**Reply by lebenski** [reply]
Oh wow adding broadcast(None) really does make a big difference!  I assume this is because without it the train still continues to broadcast it's "I'm ready for the next card" message which just clogs up the bluetooth traffic between the two hubs?

And that train motor looks brilliant!  I have one of those knocking around so will order a few more gears and get to work trying that out.  This has been such a useful conversation for me, thanks again!

**Reply by lebenski** [reply]
Hey @laurensvalk .  

Just wanted to report back to say that I took all of your advice and it's working much better!  Having the color sensor on the train just makes the whole setup a lot quicker and more responsive so I'm definitely prepared to forgo the ability to easily extend the train.  I also set up the train to run off a standard Large Technic motor (88013) and it runs a lot better now.  Fast with consistent speed and smooth acceleration.  I'm not sure if my gearing setup (or in fact the build in general) is ideal but it seems to work well enough.  I'll be looking to add a bit more weight to the train and carriages to improve overall stability but I feel like this portion of the project is 90% there now.  Here's an updated video of my setup

https://www.dropbox.com/scl/fi/6jqnriru0dh91zxum13qr/IMG_5625.MOV?rlkey=twuxcl1dsgbt6fnqjkbfrc41b&dl=0

Thanks again, Ben

**Reply by laurensvalk** [reply]
Looks awesome!

> I'm not sure if my gearing setup (or in fact the build in general) is ideal but it seems to work well enough

You may be able to get even smoother motion by gearing down instead of up (smaller gear on the motor axle).

This gives you more rotations of the motor for the same amount of train motion, so a higher resolution of millimeter train movement per degree of motor turned.

**Reply by lebenski** [reply]
Thanks.  I’m just learning the basics when it comes to gearing.  As I understand it going from a small gear to a big gear would provide more torque (stability) but less rotations of the wheels, and hence less speed.  Is that correct?

**Reply by laurensvalk** [reply]
Correct, but you can then tell the motors to drive a bit faster to keep the same overal train speed. (From the video, I think you're not at the maximum speed yet). These motors tend to be more accurate at higher speeds because of the relatively high friction at low speeds.
